### ISSUE TITLE
Better user-facing logging vs exception handling

### ISSUE THREAD

**Issue by tsjackson-noaa** [NOAA-GFDL/MDTF-diagnostics#99]
[writeup in progress] Need a better way to distinguish raising and handling of exceptions during program's normal operation from legitimate errors that the user should be notified of.

**Comment by tsjackson-noaa** [comment]
As an example, here's an annotated log of the query-fetch-preprocess section of a package run on the Travis test data. This is from e2bf979c6786eb28ee928486202b9ebf84f932f9 and line numbers correspond to those in the [Travis log](https://travis-ci.org/github/tsjackson-noaa/MDTF-diagnostics/jobs/754839694).

Here is the Varlist prior to being queried. In particular, EOF_500hPa requests zg @ 500mb (id &#35; 2). If that's not available, it will try 4D zg on pressure levels (id # 5) and if that's not available, it will use vinth2p() in NCL to interpolate the 500b level from data on hybrid sigma levels (id # 3), for which it also needs surface pressure (id # 4).
```
1441 DEBUG: Copied value of Z slice (=500 hPa) of 'plev' (identical units).
1442 DEBUG: Renaming 500 hPa Z slice of 'zg' to 'zg500'.
1443 DEBUG: ExtractLevelFunction for <#2.EOF_500hPa:zg>: add translated <zg (=geopotential_height)> as alternate for <zg500 (=geopotential_height) @ 500 hPa>
1444 DEBUG: ######################################################################
1445 DEBUG: Pre-query varlists for NCAR-CAM5.timeslice:
1446 DEBUG: <#2.EOF_500hPa:zg (=zg500) @ 1mo, 500 hPa; active:INITED, ok, required>
1447     Translation: <zg500 (=geopotential_height) @ 500 hPa>
1448     Alternate set #1: [<#5.EOF_500hPa:zg (=zg) @ 1mo, 500 hPa>]
1449     Alternate set #2: [<#3.EOF_500hPa:zg_hybrid_sigma (=zg) @ 1mo>, <#4.EOF_500hPa:ps (=ps) @ 1mo>]
1450 DEBUG: <#5.EOF_500hPa:zg (=zg) @ 1mo, 500 hPa; inactive:INITED, ok, alternate>
1451     Translation: <zg (=geopotential_height)>
1452     Alternate set #1: [<#3.EOF_500hPa:zg_hybrid_sigma (=zg) @ 1mo>, <#4.EOF_500hPa:ps (=ps) @ 1mo>]
1453 DEBUG: <#3.EOF_500hPa:zg_hybrid_sigma (=zg) @ 1mo; inactive:INITED, ok, alternate>
1454     Translation: <zg (=geopotential_height)>
1455 DEBUG: <#4.EOF_500hPa:ps (=ps) @ 1mo; inactive:INITED, ok, alternate>
1456     Translation: <ps (=surface_air_pressure)>
1457 DEBUG: <#7.example:tas (=tas) @ 1mo; active:INITED, ok, required>
1458     Translation: <tas (=air_temperature)>
1459 DEBUG: ######################################################################
```

Here's the behavior that prompted the creation of this issue. The NCAR-CAM5.timeslice run data doesn't include zg500, so that query fails. It finds zg but has no way of knowing what the pressure axis is (based on information from the filename, prior to examining the file's contents); it expects the data to be on pressure levels but finds it on sigma levels. The preprocessor can't deal with this, so it raises an error that's reported to the user.

```
1466 DEBUG: Query batch: [<#2.EOF_500hPa:zg>, <#7.example:tas>]
1467     Querying <zg500 (=geopotential_height) @ 500 hPa>
1468     No data found for <zg500 (=geopotential_height) @ 500 hPa>.
...
1475 DEBUG: Query batch: [<#5.EOF_500hPa:zg>]
1476     Querying <zg (=geopotential_height)>
1477 DEBUG: Query found <expt_key=('NCAR-CAM5.timeslice', '', ''), data_key=(1,)> for <#5.EOF_500hPa:zg>
...
1501 ERROR: Caught exception processing <#5.EOF_500hPa:zg (=zg) @ 1mo, 500 hPa> with data_key=(1,): DataPreprocessError(Data preprocessing error for data in zg: Error in read/parse data for <#5.EOF_500hPa:zg>.)
1502 Traceback (most recent call last):
1503   File "/home/travis/build/tsjackson-noaa/MDTF-diagnostics/src/preprocessor.py", line 419, in process
1504     ds = xr_util.DatasetParser().parse(ds, var)
1505   File "/home/travis/build/tsjackson-noaa/MDTF-diagnostics/src/xr_util.py", line 687, in parse
1506     self.check_variable(ds, var.translation)
1507   File "/home/travis/build/tsjackson-noaa/MDTF-diagnostics/src/xr_util.py", line 615, in check_variable
1508     self.check_names_and_units(coord, ds, ds_coord_name, update_name=True)
1509   File "/home/travis/build/tsjackson-noaa/MDTF-diagnostics/src/xr_util.py", line 562, in check_names_and_units
1510     update_our_var=True, update_ds=True)
1511   File "/home/travis/build/tsjackson-noaa/MDTF-diagnostics/src/xr_util.py", line 506, in check_attr
1512     update_our_var=update_our_var, update_ds=update_ds
1513   File "/home/travis/build/tsjackson-noaa/MDTF-diagnostics/src/xr_util.py", line 474, in _compare_attr
1514     raise TypeError((f"Found unexpected {our_attr_name} for variable "
1515 TypeError: Found unexpected standard_name for variable 'lev': 'atmosphere_hybrid_sigma_pressure_coordinate' (expected 'air_pressure').
1516
1517 The above exception was the direct cause of the following exception:
1518
1519 Traceback (most recent call last):
1520   File "/home/travis/build/tsjackson-noaa/MDTF-diagnostics/src/data_manager.py", line 637, in preprocess_data
1521     pod.preprocessor.process(v)
1522   File "/home/travis/build/tsjackson-noaa/MDTF-diagnostics/src/preprocessor.py", line 422, in process
1523     f"{var.full_name}."), var) from exc
1524 src.util.exceptions.DataPreprocessError: Data preprocessing error for data in zg: Error in read/parse data for <#5.EOF_500hPa:zg>.
1525 DEBUG: Eliminating data_key=(1,) for <#5.EOF_500hPa:zg>.
1526 DEBUG: Deactivating <#5.EOF_500hPa:zg> since all data_keys eliminated ([(1,)]).
```

The situation isn't critical for the execution of the POD, however, since the next query on the second set of alternate variables succeeds (as intended).

```
1546 DEBUG: Updating active PODs for CASENAME 'NCAR-CAM5.timeslice'
1547 DEBUG: Updating active vars for POD EOF_500hPa
1548 Request for <#5.EOF_500hPa:zg (=zg) @ 1mo, 500 hPa> failed; finding alternate vars.
1549 DEBUG: Updating active vars for POD example
1550 DEBUG: Query batch: [<#3.EOF_500hPa:zg_hybrid_sigma>, <#4.EOF_500hPa:ps>]
1551     Querying <zg (=geopotential_height)>
1552 DEBUG: Query found <expt_key=('NCAR-CAM5.timeslice', '', ''), data_key=(1,)> for <#3.EOF_500hPa:zg_hybrid_sigma>
1553     Querying <ps (=surface_air_pressure)>
1554 DEBUG: Query found <expt_key=('NCAR-CAM5.timeslice', '', ''), data_key=(0,)> for <#4.EOF_500hPa:ps>
```

One way to deal with this particular case is, upon encountering an error in the preprocessing step (ie, the file contents don't correspond to the variable we thought we were getting), the package could iterate through the remaining alternate variables and see if any of them correspond to the data it received, and then proceed on that hypothesis.

**Comment by tsjackson-noaa** [comment]
Closing this issue as resolved by PR #161.
